monomerdb/localdb/localdb.go

```diff
@@ -0,0 +1,464 @@
+	if err != nil {
+		return nil, fmt.Errorf("get header by height: %w", err)
+	}
+	txs, err := txsInRange(s, heightBytes, marshalUint64(endian.Uint64(heightBytes)+1))
+	if err != nil {
+		return nil, err
+	}
+	return monomer.NewBlock(header, txs), nil
+}
+
+func (db *DB) BlockByHash(hash common.Hash) (*monomer.Block, error) {
+	var header *monomer.Header
+	var txs bfttypes.Txs
+	if err := db.view(func(s *pebble.Snapshot) (err error) {
+		header, err = headerByHash(s, hash)
+		if err != nil {
+			return fmt.Errorf("get header by hash: %w", err)
+		}
+		txs, err = txsInRange(s, marshalUint64(uint64(header.Height)), marshalUint64(uint64(header.Height+1)))
+		if err != nil {
+			return err
+		}
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+	return monomer.NewBlock(header, txs), nil
+}
+
+func (db *DB) BlockByLabel(label eth.BlockLabel) (*monomer.Block, error) {
+	var header *monomer.Header
+	var txs bfttypes.Txs
+	if err := db.view(func(s *pebble.Snapshot) error {
+		var err error
+		header, err = headerByLabel(s, label)
+		if err != nil {
+			return fmt.Errorf("get header by label: %w", err)
+		}
+		txs, err = txsInRange(s, marshalUint64(uint64(header.Height)), marshalUint64(uint64(header.Height+1)))
+		if err != nil {
+			return err
+		}
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+	return monomer.NewBlock(header, txs), nil
+}
+
+func (db *DB) HeaderByHash(hash common.Hash) (*monomer.Header, error) {
+	var header *monomer.Header
+	if err := db.view(func(s *pebble.Snapshot) error {
+		var err error
+		header, err = headerByHash(s, hash)
+		return err
+	}); err != nil {
+		return nil, err
+	}
+	return header, nil
+}
+
+func (db *DB) HeaderByLabel(label eth.BlockLabel) (*monomer.Header, error) {
+	var header *monomer.Header
+	if err := db.view(func(s *pebble.Snapshot) error {
+		var err error
+		header, err = headerByLabel(s, label)
+		return err
+	}); err != nil {
+		return nil, err
+	}
+	return header, nil
+}
+
+func (db *DB) HeaderByHeight(height uint64) (*monomer.Header, error) {
+	return headerByHeight(db.db, marshalUint64(height))
+}
+
+func txsInRange(s *pebble.Snapshot, startHeightBytes, endHeightBytes []byte) (_ bfttypes.Txs, err error) {
+	iter, err := s.NewIter(&pebble.IterOptions{
+		LowerBound: bucketTxByHeightAndIndex.Key(startHeightBytes),
+		UpperBound: bucketTxByHeightAndIndex.Key(endHeightBytes),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("new iterator: %v", err)
+	}
+	defer func() {
+		err = wrapCloseErr(err, iter)
+	}()
+	txs := bfttypes.Txs{}
+	for iter.First(); iter.Valid(); iter.Next() {
+		value, err := iter.ValueAndErr()
+		if err != nil {
+			return nil, fmt.Errorf("get value from iterator: %v", err)
+		}
+		tx := make([]byte, len(value))
+		copy(tx, value)
+		txs = append(txs, tx)
+	}
+	return txs, nil
+}
+
+type getter interface {
+	Get([]byte) ([]byte, io.Closer, error)
+}
+
+func headerByHeight(g getter, heightBytes []byte) (_ *monomer.Header, err error) {
```

Comment:
For my own learning about go styling conventions, when should you use named return values instead of nameless return values?

Reply:
Named return values can be tricky. I only use them for wrapping errors in `defer`s. I prefer to keep all other return values declared inside the function itself.

From the go spec on [declarations](https://go.dev/ref/spec#Declarations_and_scope)
> The blank identifier may be used like any other identifier in a declaration, but it does not introduce a binding and thus is not declared.

Also see the spec for the [blank identifier](https://go.dev/ref/spec#Blank_identifier)

Reply:
This is just my habit though. I haven't seen good alternatives yet, but happy to change if someone shows me something better.